### DISCUSSION TITLE
Implement macros inside dot notation to access numeric keys

### DISCUSSION THREAD

**Discussion by VitorTS** [laravel/framework#47137] [Ideas]
In python we are able to get a reference to the last element of an array with this syntax:

`arr[-1] = "test"`

this is much better than the current PHP's syntax:

`arr[array_key_last(arr)] = "test"`

This the difference is especially noticeable when accessing the leaf nodes of a nested container (my use case for this feature):

Python:
```
animals = [
    {
       'type': 'mammals',
       'members': [
            {
                'type': 'primates',
                'members': [
                    'gorilla'
                ]
            }
       ]
    }
];

animals[-1]['members'][-1]['members'] = 'human';

```

compare to PHP:
```
$animals = [
    [
       'type'=> 'mammals',
       'members'=> [
            [
                'type'=> 'primates',
                'members'=> [
                    'gorilla'
                ]
            ]
       ]
    ]
];

$idx1 = array_key_last($animals);
$idx2 = array_key_last($animals[$idx1]['members']);
$idx3 = array_key_last($animals[$idx1]['members'][$idx2]['members']);
$animals[$idx1]['members'][$idx2]['members'][$idx3] = 'human';
```

We could have this same conveniences on Laravel by adding special macros to the dot notation:

`Arr::set($animals, '$last.members.$last.members', 'human');`

I'm not sure which syntax to use, may be this belongs on a different method or something. But those features are used a lot in the python world to simplify data manipulation and would add a lot of value to to the `Arr` helper methods.

**Comment by henzeb** [comment]
$last would be confusing and can give unwanted results for users that use or need double quotes.

**Comment by yuan8** [comment]
```
$data=[
        [
            "name"=>"Orang",
            "member"=>[
                [
                    "name"=>"C",
                    "member"=>[
                        "Babi"
                    ],
                    "foot"=>4
                ]
            ]
        ],
        [
            "name"=>"animal",
            "member"=>[
                [
                    "name"=>"A",
                    "member"=>[
                        "Babi"
                    ]
                ]
            ]
        ]
    ];
    data_fill($data,"*.member.*.foot",5)
```

this for append field for index have not field and add value default 5


for push (specific index dont use * )  array u must have specific last index +1 and user function `data_set()`

**Reply by VitorTS** [reply]
@yuan8 The idea is to get a reference to the last element. Not every element.

**Reply by Sophist-UK** [reply]
There are some difficulties with recognising -1 as meaning the last element of an array.

In Python Arrays have fixed indexes which start at 0 and are sequential integers. For named indexes you use a Dict or OrderedDict instead.

But in PHP we have an issue:
``` PHP
$ar = array("-1": "first")
$ar[] = "second"
```
In this example, what would you expect to get with `$ar[-1]`?

(Apologies if this isn't quite right because of the use of a string "-1" as a key - but you get the drift I think.)

That said, I think something similar would be possible using ArracyAccess. See https://www.php.net/manual/en/class.arrayaccess.php which shows that it should be possible to create an ArrayAccess approach which can recognise something that uniquely means "last".

So the challenge is now to create a terse way of indicating a relative index. As an example, we could implement a `Slice` class which takes 1-3 positional parameters. In your example you would then code:
``` php
$animals[Slice(-1)]['members'][Slice(-1)]['members'] = 'human';
```

Obviously a shorter class name than Slice would be preferable. "_" is already used for I18n. So you would have to see if you could find something terser. "A valid class name starts with a letter or underscore, followed by any number of letters, numbers, or underscores." So you could for example use `_A` or `A_` as a class name which is quite terse. 

If there was enough takeup, then you might even get a PHP enhancement which did something a bit like [array unpacking](https://www.php.net/manual/en/language.types.array.php#language.types.array.unpacking) which is indicated with `...$ar`.

This is just a first idea - I am sure that someone can take this further if they are passionate enough.

**Comment by shaedrich** [comment]
JavaScript had a similar problem and therefor added the [`Array.prototype.at()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at) method, so we could have a similar `Arr::at()` helper method.

**Reply by shaedrich** [reply]
Alternatively, we could have something like [XPath](https://scrapfly.io/blog/how-to-select-last-element-in-xpath/) has (like `Arr::xpath('animals[last()]["members"][last()]["members"]')`) in general

**Reply by shaedrich** [reply]
When it's always the last element of each level, there could be something like `Arr::deepAppend()` or the like 🤔

**Reply by shaedrich** [reply]
And TIL, there's also CakePHP's [hash path syntax](https://book.cakephp.org/4/en/core-libraries/hash.html#hash-path-syntax)
